### ISSUE TITLE
Applying SHAP with custom tokenizer to BERT

### ISSUE THREAD

**Issue by FiammettaC** [shap/shap#2331]
Is it correct to use SHAP + a masker with a custom tokenizer to explain a fine-tuned BERT model?

My doubt is linked to the different tokenization, since BERT tokenizer does WordPiece tokenization (e.g. 'lovely' becomes 'love', '##ly'), but with SHAP I would like to have token level explanations, so using a is a simple white space tokenizer.

Without the masker, I obtain this result:
<img width="491" alt="image" src="https://user-images.githubusercontent.com/26167801/146790543-c6f3603b-bad9-4dba-81ba-c0fa43e1d5de.png">

While, with the masker, I get the result that I want (lovely is one word, not split according to the BERT tokenization.
Here is an extract from the code, with masker and explainer:
`masker = shap.maskers.Text(r"\s")`
 `explainer = shap.Explainer(f, masker)`
 `shap_values = explainer(sentences)`

f is the prediction function, which takes the fine-tuned BERT model and the pre-trained BERT tokenizer as input.

But is this approach correct? 
Can I use a white space tokenizer, although my model is trained differently?

**Comment by umdilk** [comment]
Hello, Have you found the answer to above problem? I am facing the same issue.

**Comment by nk-fouque** [comment]
If you showed what's in f it could be clearer.
In my case I use a transformers pipeline as is shown in [this example](https://shap.readthedocs.io/en/latest/example_notebooks/text_examples/sentiment_analysis/Emotion%20classification%20multiclass%20example.html) 
Since the tokenizer is inside the pipeline, I think putting an other "tokenizer" in the masker will only pre-split the data for masking purposes.

This is just speculation but since running it with no given masker and running it with this custom masker give the same f_label(x), and only affect the explanation part, I think the approach is correct.

**Comment by TomatenMarc** [comment]
Hey, I had the same issue.

I solved this problem like this:

```python 
masker = shap.maskers.Text(tokenizer=r"\W+") # this will create a basic whitespace tokenizer
explainer= shap.Explainer(f, masker) #f is a classification pipeline
```
